async-nats/src/service/mod.rs

```diff
@@ -633,61 +473,220 @@ impl Service {
     }
 
     /// Returns [Stats] for this service instance.
-    pub async fn stats(&self) -> Stats {
-        self.stats.lock().unwrap().clone()
+    pub async fn stats(&self) -> HashMap<String, endpoint::Stats> {
+        self.endpoints_state
+            .lock()
+            .unwrap()
+            .endpoints
+            .iter()
+            .map(|(key, value)| (key.to_owned(), value.to_owned().into()))
+            .collect()
     }
 
     /// Returns [Info] for this service instance.
     pub async fn info(&self) -> Info {
         self.info.clone()
     }
 
+    /// Creates a group for endpoints under common prefix prefix.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # #[tokio::main]
+    /// # async fn main() -> Result<(), async_nats::Error> {
+    /// use async_nats::service::ServiceExt;
+    /// let client = async_nats::connect("demo.nats.io").await?;
+    /// let mut service = client.service_builder().start("service", "1.0.0").await?;
+    ///
+    /// let v1 = service.group("v1");
+    /// let products = v1.endpoint("products").await?;
+    /// # Ok(())
+    /// # }
+    /// ```
     pub fn group<S: ToString>(&self, prefix: S) -> Group {
         Group {
+            subjects: self.subjects.clone(),
             prefix: prefix.to_string(),
-            stats: self.stats.clone(),
+            stats: self.endpoints_state.clone(),
             client: self.client.clone(),
             shutdown_tx: self.shutdown_tx.clone(),
         }
     }
+
+    /// Builder for customized [Endpoint] creation.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # #[tokio::main]
+    /// # async fn main() -> Result<(), async_nats::Error> {
+    /// use async_nats::service::ServiceExt;
+    /// let client = async_nats::connect("demo.nats.io").await?;
+    /// let mut service = client.service_builder().start("service", "1.0.0").await?;
+    ///
+    /// let products = service
+    ///     .endpoint_builder()
+    ///     .name("api")
+    ///     .add("products")
+    ///     .await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn endpoint_builder(&self) -> EndpointBuilder {
```

Comment:
Bikeshedding:
endpoint_builder vs build_endpoint.

```suggestion
    pub fn endpoint_builder(&self) -> EndpointBuilder {
```

Reply:
Disagree. `build_endpoint` sounds like finalizer.